[Tidy] Store `page.id` in outer page container to enable page specific styling (#455)
mckinsey/vizro: vizro-core/docs/pages/user-guides/assets.md

    [AssetsCSS]: ../../assets/user_guides/assets/css_change.png

## Overwrite CSS properties for specific pages

> To style components for only a specific page, use the page's `id` in CSS selectors. By default, this is the [same as the page `title`](pages.md), but such a value might not be a valid CSS identifier. A suitable `id` must be unique across all models in the dashboard and should contain only alphanumeric characters, hyphens (`-`) and underscores (`_`). In particular, note that spaces are _not_ allowed.

Reviewer: ```suggestion
To style components for a specific page, use the page's `id` in CSS selectors. By default, this is the [same as the page `title`](pages.md), but such a value might not be a valid CSS identifier. A suitable `id` must be unique across all models in the dashboard and should contain only alphanumeric characters, hyphens (`-`) and underscores (`_`). In particular, note that spaces are _not_ allowed.
```

    [AssetsCSS]: ../../assets/user_guides/assets/css_change.png

## Overwrite CSS properties for specific pages

> If you want to style specific pages only, you need to provide a valid `id` to the page you want to modify.
> This unique `id` enables you to use CSS selectors to target specific pages and their components for styling.
> 
> An HTML `id` is considered valid if the following conditions are met:
> - Use lowercase letters and hyphens
> - Avoid using special characters or spaces
> - Provide a unique `id`

Reviewer 2: ```suggestion
To style components for only a specific page, use the page's `id` in CSS selectors. By default, this is the [same as the page `title`](pages.md), but such a value might not be a valid CSS identifier. A suitable `id` must be unique across all models in the dashboard and should contain only alphanumeric characters, hyphens (`-`) and underscores (`_`). In particular, note that spaces are _not_ allowed.
```

This guidance is general about `id` really applies everywhere of course, not just to page's. Feel free to move it now or in the future when we restructure the page as discussed earlier.

The only special part that applies specifically to page is just that the `id` is automatically set and hence more likely to be inappropriate for CSS identifiers due to having a space in it.

Suppose you want to hide the page title on one page only. Here's how you can achieve this:

> 1. Provide a valid `id` to the `Page`, for example `Page(id="page-with-hidden-title", title="Page with hidden title", ...)`
> 2. Identify the CSS class or ID you need to target. To hide the page title, you need to hide the container with the ID `right-header`.
> 3. Use the `id` in combination with CSS selectors to change the relevant CSS properties
> 4. Add your custom css file to the `assets` folder as explained above

Reviewer 2: ```suggestion
1. Provide a valid `id` to the `Page`, for example `Page(id="page-with-hidden-title", title="Page with hidden title", ...)`.
2. Identify the CSS class or ID you need to target. To hide the page title, you need to hide the container with the ID `right-header`.
3. Use the `id` in combination with CSS selectors to change the relevant CSS properties.
4. Add your custom css file to the `assets` folder as explained above.
```

    [AssetsCSS]: ../../assets/user_guides/assets/css_change.png

## Overwrite CSS properties for specific pages
Each page is assigned a unique identifier (`page.id`), which is sanitized and then stored as the ID of an outer `<div>` element.
This identifier is formatted according to CSS naming conventions by removing non-alphanumeric characters and replacing spaces with dashes.

> For instance, if a `page.id` is "Variable Analysis," it will be stored as a CSS ID of "variable-analysis."

Reviewer 2: ```suggestion
For instance, if a `page.id` is `Variable Analysis` then it has CSS ID `variable-analysis`.
```

As it stands the "," and "." are very confusing here because it looks like you're replacing "," with ".".

Not sure whether we should convert to lowercase, but let's discuss 🤔

Author: Oh, I don't even know how this got in - probably my automatic grammar checker😆 There is not supposed to be a comma 😄

    === "my_css_file.css"
    ```css
    #page-with-hidden-title #right-header {
      display: none;
    }
    ```
    === "app.py"
        ```py
        import vizro.models as vm
        from vizro import Vizro

        page_one = vm.Page(
            title="Page with hidden title",
            components=[vm.Card(text="""# Placeholder""")]
        )

        page_two = vm.Page(
            title="Page with shown title",
            components=[vm.Card(text="""# Placeholder""")]
        )

        dashboard = vm.Dashboard(pages=[page_one, page_two])
        Vizro().build(dashboard).run()
        ```
    === "app.yaml"
        ```yaml
        # Still requires a .py to add data to the data manager and parse YAML configuration
        # See yaml_version example
        pages:
        - components:
            - text: |
                # Placeholder
              type: card
          title: Page with hidden title
        - components:
            - text: |
                # Placeholder
              type: card
          title: Page with shown title
        ```
    === "Result"
         [![PageTitle]][PageTitle]

    [PageTitle]: ../../assets/user_guides/assets/css_page_title.png

> ## Overwrite CSS properties in selective components

Reviewer 2: Is this section still correct? e.g. it talks a lot about `_outer`, and I'm not sure whether that's still used much after your recent bootstrap refactorings?

Author: This still holds, but it gets a bit confusing now because it does not hold for all components. Let's discuss in our call, but just to show you what I mean:

Here we don't have the `_outer` suffix for an ID, because we can provide the self.id directly to the component that gets outputted:

`dbc.Button(id=self.id, children=self.text)`

However, here we still need the `_outer` suffix because of two reasons:

1. We once aligned that the `self.id` is being provided to the inner component that might be reactive to callbacks. In the example of the card it is the `dcc.Markdown` component. Because of that we cannot provide the `self.id` to the outermost component. 
2. Let's say someone wants to change the card background-color but only for a selective card and not for all. Here we need to provide an `id` to the outermost wrapper of the card so we can selectively target all inner components. And because we can't use `self.id`, we started using `self.id_outer`.

Happy to change this current system though, just not sure what's the best approach here 🤔 

`dbc.Card(card_content, className=card_class, id=f"{self.id}_outer")`

Reviewer 2: Yeah, that is what I expected but I didn't know how prevalent the `_outer` suffix was any more.

I think the current system basically makes sense (using `self.id` for the "core" reactive component rather than the outer container) but we should _consider_ (but maybe not do...) removing the `_outer` ones if they're just there for CSS selectors [in favour of using `has()`](https://stackoverflow.com/questions/1014861/is-there-a-css-parent-selector).

Author: Yes, I like that! Let's discuss in our call later. But I would also be in favour of killing the entire _outer system and just updating the docs 👍
